plugins/Files/js/sagas/files.js

```diff
@@ -0,0 +1,227 @@
+import Path from 'path'
+import * as actions from '../actions/files.js'
+import * as constants from '../constants/files.js'
+import { List } from 'immutable'
+import { sendError, siadCall, totalUsage, readdirRecursive, parseDownloads, parseUploads, estimatedStoragePriceGBSC } from './helpers.js'
+
+const blockMonth = 4382
+const allowanceMonths = 3
+const allowanceHosts = 24
+const allowancePeriod = blockMonth*allowanceMonths
+
+// Query siad for the state of the wallet.
+// dispatch `unlocked` in receiveWalletLockstate
+function* getWalletLockstateSaga() {
+	try {
+		const response = yield siadCall('/wallet')
+		yield put(actions.receiveWalletLockstate(response.unlocked))
+	} catch (e) {
+		sendError(e)
+	}
+}
+
+// Query siad for the user's files.
+function* getFilesSaga() {
+	try {
+		const response = yield siadCall('/renter/files')
+		const files = List(response.files)
+		yield put(actions.receiveFiles(files))
+		yield put(actions.receiveDiskUsage(totalUsage(files)))
+	} catch (e) {
+		sendError(e)
+	}
+}
+
+// Set the user's renter allowance.
+function* setAllowanceSaga(action) {
+	try {
+		const response = yield siadCall('/renter')
+		const newAllowance = SiaAPI.siacoinsToHastings(action.funds).add(response.settings.allowance.funds)
+		yield siadCall({
+			url: '/renter',
+			method: 'POST',
+			qs: {
+				funds: newAllowance.toString(),
+				hosts: allowanceHosts,
+				period: allowancePeriod,
+			},
+		})
+		yield put(actions.setAllowanceCompleted())
+		yield put(actions.closeAllowanceDialog())
+	} catch (e) {
+		yield put(actions.setAllowanceCompleted())
+		yield put(actions.closeAllowanceDialog())
+		sendError(e)
+	}
+}
+
+// Calculate monthly storage cost from a requested monthly storage amount
+function* calculateStorageCostSaga(action) {
+	try {
+		const response = yield siadCall('/renter/hosts/active')
+		const cost = estimatedStoragePriceGBSC(response.hosts, action.size, allowancePeriod)
+		yield put(actions.setStorageCost(cost.round(3).toString()))
+		yield put(actions.setStorageSize(action.size))
+	} catch (e) {
+		console.error(e)
+		yield put(actions.setStorageSize(''))
+		yield put(actions.setStorageCost('0'))
+	}
+}
+
+// Query Siad for the current wallet balance.
+function* getWalletBalanceSaga() {
+	try {
+		const response = yield siadCall('/wallet')
+		const confirmedBalance = SiaAPI.hastingsToSiacoins(response.confirmedsiacoinbalance).round(2).toString()
+		yield put(actions.receiveWalletBalance(confirmedBalance))
+	} catch (e) {
+		sendError(e)
+	}
+}
+
+const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms))
+
+function* setAllowanceProgressBarSaga() {
+	try {
+		let response = yield siadCall('/renter/contracts')
+		if (!response.contracts) {
+			return
+		}
+		const initialContracts = response.contracts.length
+		while (true) {
+			response = yield siadCall('/renter/contracts')
+			const deltaContracts = response.contracts.length - initialContracts
+			if (deltaContracts === allowanceHosts) {
```

Comment:
so this will only exit when an additional 24 contracts are formed? (as opposed to 24 total)
hmm. Not sure if that's the correct behavior.

At the very least, in these circumstances you should generally avoid `==` in favor of `>=`